File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/ClusterableMarkerHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.Marker
import com.naver.maps.map.overlay.Overlay
import io.flutter.plugin.common.MethodChannel

internal interface ClusterableMarkerHandler : MarkerHandler, ClusterMarkerHandler {
    fun handleClusterableMarker(
        marker: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (marker as Marker).let { m ->
        val isClusterMarkerMode = method.startsWith("l")
        if (isClusterMarkerMode) {
            return handleClusterMarker(m, method, arg, result)
        }

        when (method) {
            else -> handleMarker(m, method, arg, result)
        }
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/ClusterableMarkerHandler.swift
import NMapsMap
import Flutter

internal protocol ClusterableMarkerHandler: MarkerHandler, ClusterMarkerHandler {
    
}

internal extension ClusterableMarkerHandler {
    func handleClusterableMarker(
        marker: NMFOverlay,
        method: String,
        arg: Any?,
        result: @escaping FlutterResult)
    {
        let marker = marker as! NMFMarker
        let isMarkerClusterMode = method.starts(with: "l")
        
        if isMarkerClusterMode {
            return handleClusterMarker(marker: marker, method: method, arg: arg, result: result)
        }
        
        switch method {
        default: handleMarker(marker: marker, method: method, args: arg, result: result)
        }
    }
}
